// master_mapper.proto

syntax = "proto3";

message Point {
  double x = 1;
  double y = 2;
}

message DataPoint {
  int32 centroid_id = 1;
  double x = 2;
  double y = 3;
}

message MapRequest {
  repeated int32 start_index = 1;  // Indexes of lines to be processed by the mapper
  repeated int32 end_index = 2;
  int32 num_reducers = 3;
  repeated Point centroids = 4;
  bool is_retry = 5;
}

message MapResponse {
  bool success = 1;  // Indicates whether mapping was successful or not
}

message ReturnDataRequest {
  int32 reducer_id = 1;
}

message ReturnDataResponse {
  repeated DataPoint data_points = 1;
}

message StartReduceRequest {
  repeated int32 mapper_ids = 1;
  int32 reducer_id = 2;
  bool is_retry = 3;
}

message StartReduceResponse {
  repeated DataPoint updated_centroids = 1;
  bool success = 2;
}



service Mapper {
  rpc Map(MapRequest) returns (MapResponse) {}
  rpc ReturnData(ReturnDataRequest) returns (ReturnDataResponse) {}
}

service Reducer {
  rpc StartReduce(StartReduceRequest) returns (StartReduceResponse) {}
}